core/src/main/java/com/linecorp/armeria/internal/common/TimeoutScheduler.java

```diff
@@ -18,73 +18,194 @@
 import com.google.common.math.LongMath;
 
 import com.linecorp.armeria.common.util.TimeoutMode;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
 
 import io.netty.channel.EventLoop;
+import io.netty.util.concurrent.EventExecutor;
 
 public final class TimeoutScheduler {
 
+    private static final AtomicReferenceFieldUpdater<TimeoutScheduler, TimeoutFuture>
+            whenTimingOutUpdater = AtomicReferenceFieldUpdater.newUpdater(
+            TimeoutScheduler.class, TimeoutFuture.class, "whenTimingOut");
+
+    private static final AtomicReferenceFieldUpdater<TimeoutScheduler, TimeoutFuture>
+            whenTimedOutUpdater = AtomicReferenceFieldUpdater.newUpdater(
+            TimeoutScheduler.class, TimeoutFuture.class, "whenTimedOut");
+
+    private static final AtomicReferenceFieldUpdater<TimeoutScheduler, Runnable>
+            pendingTimeoutTaskUpdater = AtomicReferenceFieldUpdater.newUpdater(
+            TimeoutScheduler.class, Runnable.class, "pendingTimeoutTask");
+
+    private static final AtomicLongFieldUpdater<TimeoutScheduler> pendingTimeoutNanosUpdater =
+            AtomicLongFieldUpdater.newUpdater(TimeoutScheduler.class, "pendingTimeoutNanos");
+
+    private static final TimeoutFuture COMPLETED_FUTURE;
+
+    static {
+        COMPLETED_FUTURE = new TimeoutFuture();
+        COMPLETED_FUTURE.doComplete();
+    }
+
+    enum State {
+        INIT,
+        INACTIVE,
+        SCHEDULED,
+        TIMED_OUT
+    }
+
     private long timeoutNanos;
+    private long firstExecutionTimeNanos;
+
+    private State state = State.INIT;
+
+    @Nullable
+    private TimeoutTask timeoutTask;
+    @Nullable
+    private ScheduledFuture<?> timeoutFuture;
+    @Nullable
+    private EventExecutor eventLoop;
+
+    // Updated via whenTimingOutUpdater
     @Nullable
-    private Consumer<TimeoutController> pendingTimeoutTask;
+    private volatile TimeoutFuture whenTimingOut;
+    // Updated via whenTimedOutUpdater
     @Nullable
-    private EventLoop eventLoop;
+    private volatile TimeoutFuture whenTimedOut;
+    // Updated via pendingTimeoutTaskUpdater
     @Nullable
-    private TimeoutController timeoutController;
+    private volatile Runnable pendingTimeoutTask;
+    // Updated via pendingTimeoutNanosUpdater
+    @SuppressWarnings("FieldMayBeFinal")
+    private volatile long pendingTimeoutNanos;
+
+    private volatile boolean initialized;
 
     public TimeoutScheduler(long timeoutNanos) {
         this.timeoutNanos = timeoutNanos;
+        pendingTimeoutNanos = timeoutNanos;
+    }
+
+    /**
+     * Initializes this {@link TimeoutScheduler}.
+     * Note that this method should be called in the specified {@link EventLoop}
+     * to initialize this scheduler synchronously
+     */
+    public void init(EventExecutor eventLoop, TimeoutTask timeoutTask, long initialTimeoutNanos) {
+        if (!eventLoop.inEventLoop()) {
+            eventLoop.execute(() -> init(eventLoop, timeoutTask, initialTimeoutNanos));
```

Comment:
Should return here?
Also we can remove the second line in Javadoc?

Reply:
Oops. It should return here.